README.md

```diff
@@ -11,130 +12,128 @@ The framework will help you to:
-a subclass of `CharmBase` and pass that into the framework's `main` function:
+> 🛈 once `charmcraft build` works the layout will get simpler!
 
-```python
-import sys
-sys.path.append('lib')  # noqa: E402
+`src/charm.py` here is the entry point to your charm code. At a minimum, it
+needs to define a subclass of `CharmBase` and pass that into the framework's
+`main` function:
 
+```python
+#!../env/bin/python3
 from ops.charm import CharmBase
 from ops.main import main
 
-
-class MyCharm(CharmBase):
-    pass
-
-
-if __name__ == "__main__":
-    main(MyCharm)
-```
-
-This charm does nothing, because the `MyCharm` class passed to the operator
-framework's `main` function is empty. Functionality can be added to the charm
-by instructing it to observe particular Juju events when the `MyCharm` object
-is initialized. For example,
-
-```python
 class MyCharm(CharmBase):
     def __init__(self, *args):
         super().__init__(*args)
         self.framework.observe(self.on.start, self.on_start)
 
     def on_start(self, event):
         # Handle the start event here.
+
+if __name__ == "__main__":
+    main(MyCharm)
 ```
 
-Every standard event in Juju may be observed that way, and you can also easily
-define your own events in your custom types.
+That should be enough for you to be able to run
+
+```
+$ juju deploy .
+```
 
-> The second argument to `observe` can be either the handler as a bound
-> method, or the observer itself if the handler is a method of the observer
-> that follows the conventional naming pattern. That is, in this case, we
-> could have called just `self.framework.observe(self.on.start, self)`.
+Happy charming!
 
-The `hooks/` directory must contain a symlink to your `src/charm.py` entry
-point so that Juju can call it. You only need to set up the `hooks/install` link
-(`hooks/start` for K8s charms, until [lp#1854635](https://bugs.launchpad.net/juju/+bug/1854635)
-is resolved), and the framework will create all others at runtime.
+## Testing your charms
 
-Once your charm is ready, upload it to the charm store and deploy it as
-normal with:
+The operator framework provides a testing harness, so that you can test that
+your charm does the right thing when presented with different scenarios, without
+having to have a full deployment to do so. `pydoc3 ops.testing` has the details
+for that, including this example:
 
-```
-# Replace ${CHARM} with the name of the charm.
-charm push . cs:~${USER}/${CHARM}
-# Replace ${VERSION} with the version created by `charm push`.
-charm release cs:~${USER}/${CHARM}-${VERSION}
-charm grant cs:~${USER}/${CHARM}-${VERSION} everyone
-# And now deploy your charm.
-juju deploy cs:~${USER}/$CHARM
+```python
+harness = Harness(MyCharm)
+# Do initial setup here
+relation_id = harness.add_relation('db', 'postgresql')
+# Now instantiate the charm to see events as the model changes
+harness.begin()
+harness.add_relation_unit(relation_id, 'postgresql/0', remote_unit_data={'key': 'val'})
```

Comment:
this is no longer correct, because we got rid of 'remote_unit_data'
```
harness.add_relation_unit(relation_id, 'postgresql/0')
harness.update_relation_data(relation_id, 'postgresql/0', {'key': 'val'})
```

Reply:
whoops :-)